stdlib/universe/narrow_state_tracking.go

```diff
@@ -0,0 +1,171 @@
+package universe
+
+import (
+	"github.com/apache/arrow/go/arrow/memory"
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/array"
+	"github.com/influxdata/flux/arrow"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/execute/table"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/influxdata/flux/values"
+)
+
+func NewNarrowStateTrackingTransformation(t *stateTrackingTransformation, id execute.DatasetID, mem memory.Allocator) (execute.Transformation, execute.Dataset, error) {
+	a := &narrowStateTrackingTransformationAdapter{t: t}
+	nt, d, err := execute.NewNarrowStateTransformation(id, a, mem)
+	if err != nil {
+		return nil, nil, err
+	}
+	return nt, d, nil
+}
+
+type narrowStateTrackingTransformationAdapter struct {
+	t *stateTrackingTransformation
+}
+
+type trackedState struct {
+	start,
+	prevTime values.Time
+
+	countInState,
+	durationInState bool
+
+	count,
+	duration int64
+}
+
+func (a *narrowStateTrackingTransformationAdapter) Process(chunk table.Chunk, state interface{}, d *execute.TransportDataset, mem memory.Allocator) (interface{}, bool, error) {
+	// Track whether or not the state has been modified
+	mod := false
+
+	if chunk.Len() == 0 {
+		return state, mod, nil
+	}
+
+	if state == nil {
+		state = trackedState{
+			count:    -1,
+			duration: -1,
+		}
+		mod = true
+	}
+	s := state.(trackedState)
+
+	cols := chunk.Cols()
+	fn, err := a.t.fn.Prepare(cols)
+	if err != nil {
+		return s, mod, err
+	}
+
+	// Get the index for the time column
+	timeIdx := execute.ColIdx(a.t.timeCol, cols)
```

Comment:
`table.Chunk` has a [dedicated method](https://pkg.go.dev/github.com/influxdata/flux/execute/table#Chunk.Index) for this.